File: finalFrontendApplication/src/app/features/devices/controllers/device-controller.ts
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { HttpClient } from "@angular/common/http";
import { Device } from "../entities/Device";



@Injectable({
    providedIn: 'root'
})
export class DeviceController {

    private url = "http://localhost:8080/devices";

    constructor(private http: HttpClient) {}



    getDevices(){
        return this.http.get<Device[]>(this.url);
    }

    getDeviceBySerialNumber(serialNumber: string): Observable<Device> {
        return this.http.get<Device>(`${this.url}/${serialNumber}`);
    }

    createDevice(device: Device): Observable<Device> {
        return this.http.post<Device>(this.url, device);
    }
    

    updateDevice(serialNumber: string, device: Device): Observable<Device> {
        return this.http.put<Device>(`${this.url}/${serialNumber}`, device);
    }

    deleteDevice(serialNumber: string): Observable<void> {
        return this.http.delete<void>(`${this.url}/${serialNumber}`);
    }



    getAllCompanyDevices(companyId: string): Observable<Device[]> {
        return this.http.get<Device[]>(`${this.url}/company/${companyId}`);

    }

    getAllEmployeeDevices(employeeId: string): Observable<Device[]> {
        return this.http.get<Device[]>(`${this.url}/employee/${employeeId}`);
    }


    assignDeviceToCompany(serialNumber: string, companyId: string): Observable<Device> {
        return this.http.put<Device>(`${this.url}/${serialNumber}/assign/company/${companyId}`, {});

    }

    assignDeviceToEmployee(serialNumber: string, employeeId: string): Observable<void> {
        return this.http.post<void>(`${this.url}/assign/${serialNumber}/employee/${employeeId}`, {});

    }
    unassignDeviceFromEmployee(serialNumber: string, employeeId: string): Observable<void> {
        return this.http.post<void>(`${this.url}/unassign/${serialNumber}/employee/${employeeId}`, {});
    }

    unassignDeviceFromCompany(serialNumber: string, companyId: string): Observable<Device> {
        return this.http.put<Device>(`${this.url}/unassign/${serialNumber}/company/${companyId}`, {});
    }

    getAllDevicesAssignedToEmployee(employeeId: string): Observable<Device[]> {
        return this.http.get<Device[]>(`${this.url}/employee/${employeeId}`);

    }

    getAllDevicesAssignedToCompany(companyId: string): Observable<Device[]> {
        return this.http.get<Device[]>(`${this.url}/company/${companyId}`);
    }

    unassignAllDevicesFromEmployee(employeeId: string): Observable<void> {
        return this.http.put<void>(`${this.url}/employee/${employeeId}/unassignAllDevices`, {});
    }

    unassignAllDevicesFromCompany(companyId: string): Observable<void> {
        return this.http.put<void>(`${this.url}/company/${companyId}/unassignAllDevices`, {});
    }

}

File: Android/app/src/main/java/com/bivolaris/myapplication/controller/DeviceController.java
package com.bivolaris.myapplication.controller;

import com.bivolaris.myapplication.model.CreateDeviceRequest;
import com.bivolaris.myapplication.model.Device;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface DeviceController {


    @GET("devices")
    Call<List<Device>> getDevices();

    @GET("devices/{serialNumber}")
    Call<Device> getDevice(@Path("serialNumber") String serialNumber);

    @POST("devices")
    Call<Void> createDevice(@Body CreateDeviceRequest device);

    @PUT("devices/{serialNumber}")
    Call<Device> updateDevice(@Path("serialNumber") String serialNumber,@Body Device device);

    @DELETE("devices/{serialNumber}")
    Call<Void> deleteDevice(@Path("serialNumber") String serialNumber);

    @POST("devices/assign/{serialNumber}/company/{companyId}")
    Call<Void> assignDeviceToCompany(@Path("serialNumber") String serialNumber,@Path("companyId") String companyId);

    @POST("devices/assign/{serialNumber}/employee/{employeeId}")
    Call<Void> assignDeviceToEmployee(@Path("serialNumber") String serialNumber,@Path("employeeId") String employeeId);

    @POST("devices/unassign/{serialNumber}/company/{companyId}")
    Call<Void> unassignDeviceFromCompany(@Path("serialNumber") String serialNumber,@Path("companyId") String companyId);

    @POST("devices/unassign/{serialNumber}/employee/{employeeId}")
    Call<Void> unassignDeviceFromEmployee(@Path("serialNumber") String serialNumber,@Path("employeeId") String employeeId);



    @GET("devices/company/{companyId}")
    Call<List<Device>> getDevicesByCompanyId(@Path("companyId") String companyId);

}
